File: src/main/java/com/chethan/assignment5/student/StudentDatabase.java
package com.chethan.assignment5.student;


import java.sql.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Supplier;
import java.util.stream.Collectors;

/**
 * Created by Chethan on Feb 24, 2022.
 */

public class StudentDatabase implements IStudentDatabase {
    private static Connection con = null;
    ExecutorService threadPool = Executors.newCachedThreadPool();

    // connecting to database
    private static void getDatabaseConnection() {
        try {
            Class.forName("org.postgresql.Driver");
            con = DriverManager.getConnection("jdbc:postgresql://localhost:5432/training?user=postgres&password=admin");
        } catch (ClassNotFoundException | SQLException e) {
            e.printStackTrace();
        }
    }

    // closing all connection with database
    private static void closeDatabaseConnection() {
        if (con != null) {
            try {
                con.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
    }

    // get all students details from database
    @Override
    public CompletableFuture<List<Student>> getStudentsData() {
        Supplier<List<Student>> studentList = () -> {
            getDatabaseConnection();

            List<Student> studentsList = new ArrayList<Student>();

            PreparedStatement pstmt = null;
            ResultSet         rs    = null;
            String            qry   = "select * from public.\"Student\"";

            try {
                pstmt = con.prepareStatement(qry);
                rs = pstmt.executeQuery();

                while (rs.next()) {
                    String       id         = rs.getString("id");
                    String       name       = rs.getString(2);
                    double       gpa        = rs.getDouble(3);
                    String       gender     = rs.getString(4);
                    int          gradeLevel = rs.getInt(5);
                    List<String> activities = Arrays.asList((String[]) rs.getArray(6).getArray());

                    studentsList.add(new Student(id, name, gpa, gender, gradeLevel, activities));

                }

            } catch (SQLException e) {
                e.printStackTrace();
            } finally {
                if (rs != null) {
                    try {
                        rs.close();
                    } catch (SQLException e) {
                        e.printStackTrace();
                    }
                }

                if (pstmt != null) {
                    try {
                        pstmt.close();
                    } catch (SQLException e) {
                        e.printStackTrace();
                    }
                }
                closeDatabaseConnection();
            }
            return studentsList;
        };
        return CompletableFuture.supplyAsync(studentList, threadPool);
    }

    //   get a student detail from database
    @Override
    public CompletableFuture<Student> getStudent(String sid) {

        Supplier<Student> student = () -> {

            getDatabaseConnection();

            Student studentObject = null;

            PreparedStatement pstmt = null;
            ResultSet         rs    = null;
            String            qry   = "select * from public.\"Student\" where \"id\"=?";

            try {
                pstmt = con.prepareStatement(qry);
                pstmt.setString(1, sid);
                rs = pstmt.executeQuery();
                if (rs.next()) {
                    String       id         = rs.getString("id");
                    String       name       = rs.getString(2);
                    double       gpa        = rs.getDouble(3);
                    String       gender     = rs.getString(4);
                    int          gradeLevel = rs.getInt(5);
                    List<String> activities = Arrays.asList((String[]) rs.getArray(6).getArray());

                    studentObject = new Student(id, name, gpa, gender, gradeLevel, activities);
                }

            } catch (SQLException e) {
                e.printStackTrace();
            } finally {
                if (rs != null) {
                    try {
                        rs.close();
                    } catch (SQLException e) {
                        e.printStackTrace();
                    }
                }

                if (pstmt != null) {
                    try {
                        pstmt.close();
                    } catch (SQLException e) {
                        e.printStackTrace();
                    }
                }

                closeDatabaseConnection();
            }
            return studentObject;
        };
        return CompletableFuture.supplyAsync(student, threadPool);
    }


    // add a new student to database
    @Override
    public CompletableFuture<Student> addStudent(Student student) {

        getDatabaseConnection();

        Statement stmt = null;

        String activities = student.getActivities().stream().map(m -> "'" + m + "'").collect(Collectors.joining(",", "Array[", "]"));

        try {
            stmt = con.createStatement();
            stmt.executeUpdate("insert into public.\"Student\" values ('" + student.getId() + "', '" + student.getName() + "', " + student.getGpa()
                    + ", '" + student.getGender() + "', " + student.getGradeLevel() + ", " + activities + ")");

        } catch (SQLException e) {
            e.printStackTrace();
        } finally {
            if (stmt != null) {
                try {
                    stmt.close();
                } catch (SQLException e) {
                    e.printStackTrace();
                }
            }

        }
        closeDatabaseConnection();
        return getStudent(student.getId());
    }

    // update student details
    @Override
    public CompletableFuture<Student> updateStudentDetails(String sid, String columnName, String value) {

        getDatabaseConnection();
        Statement stmt = null;
        try {
            stmt = con.createStatement();
            stmt.executeUpdate("update public.\"Student\" set \"" + columnName + "\"=" + value + " where \"id\"='" + sid + "'");
        } catch (SQLException e) {
            e.printStackTrace();
        } finally {
            if (stmt != null) {
                try {
                    stmt.close();
                } catch (SQLException e) {
                    e.printStackTrace();
                }
            }

        }
        closeDatabaseConnection();
        return getStudent(sid);
    }

    // delete student by ID
    @Override
    public void deleteStudentsById(String sid) {
        getDatabaseConnection();
        Statement stmt = null;
        try {
            stmt = con.createStatement();
            stmt.executeUpdate("delete from public.\"Student\" where \"id\"='" + sid + "'");
        } catch (SQLException e) {
            e.printStackTrace();
        } finally {
            if (stmt != null) {
                try {
                    stmt.close();
                } catch (SQLException e) {
                    e.printStackTrace();
                }
            }
        }
        closeDatabaseConnection();
    }
}

File: src/main/scala/com/chethan/assignment1/StudentDatabase.scala
package com.chethan.assignment1

import java.sql.{Connection, DriverManager, ResultSet}

import scala.concurrent.Future
import scala.util.{Failure, Success, Try}

/**
  * Created by Chethan on Aug 01, 2022.
  */
case class Student(id: String, name: String, gpa: Double, gender: String, gradeLevel: Int, activities: List[String]) extends SuperStudents {

  def isMale = gender == "male"

  def isFemale = !isMale

}

class StudentDatabase extends DataBase[Student] {

  private def getConnection(host: String, username: String, password: String): Connection = {
    val driver  = "org.postgresql.Driver"
    val jdbcUrl = s"jdbc:postgresql://$host:5432/training?user=$username&password=$password"
    Class forName driver
    DriverManager getConnection jdbcUrl
  }

  override def getAllData: Future[List[Student]] = {
    val getAllQuery = s""" SELECT * FROM "public"."Student" """

    def getAllFromResultSet(resultSet: ResultSet): List[Student] = {
      def resultSetLoop(resultSetData: ResultSet, listOfStudents: List[Student]): List[Student] = {
        if (!resultSet.next())
          listOfStudents
        else {
          val id                       = resultSet getString "id"
          val name                     = resultSet getString 2
          val gpa                      = resultSet getDouble 3
          val gender                   = resultSet getString 4
          val gradeLevel               = resultSet getInt 5
          val rawSqlActivity           = resultSet.getArray(6).toString
          val activities: List[String] = rawSqlActivity.substring(1, rawSqlActivity.length - 1).split(",").toList
          resultSetLoop(resultSetData, Student(id, name, gpa, gender, gradeLevel, activities) +: listOfStudents)
        }
      }


//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= 24242424242424242424 happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= 24242424242424242424 happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= 24242424224424242424 happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= 24242424242244242424 happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= 24242424242244242424 happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= 24242424242424242424 happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"
//      val val val val val  happyIdependenceDay= happyIdependenceDay= happyIdependenceDay= "happyIndependenceday"




      resultSetLoop(resultSet, Nil)
    }

    val result = Try[List[Student]]{
      val connection     = new StudentDatabase() getConnection("127.0.0.1", "postgres", "admin")
      val statement      = connection createStatement()
      val resultSet      = statement executeQuery getAllQuery
      val listOfStudents = getAllFromResultSet(resultSet)
      resultSet close()
      statement close()
      connection close()
      listOfStudents
    }

    result match {
      case Success(value)     => Future successful value
      case Failure(exception) => Future failed exception
    }
  }


  override def getDataById(id: AnyVal): Future[Student] = {
    val getDataByIdQuery = s"""SELECT * FROM "public"."Student" WHERE id = '$id' """

    val result = Try[Student]{
      val connection = new StudentDatabase() getConnection("127.0.0.1", "postgres", "admin")
      val statement  = connection createStatement()
      val resultSet  = statement executeQuery getDataByIdQuery
      resultSet next()
      val id                       = resultSet getString "id"
      val name                     = resultSet getString 2
      val gpa                      = resultSet getDouble 3
      val gender                   = resultSet getString 4
      val gradeLevel               = resultSet getInt 5
      val rawSqlActivity           = resultSet.getArray(6).toString
      val activities: List[String] = rawSqlActivity.substring(1, rawSqlActivity.length - 1).split(",").toList
      resultSet close()
      statement close()
      connection close()
      Student(id, name, gpa, gender, gradeLevel, activities)
    }
    result match {
      case Success(value)     => Future successful value
      case Failure(exception) => Future failed exception
    }
  }

  override def addData(student: Student): Future[Boolean] = {
    val activities     = student.activities.mkString("Array['", "','", "']")
    val getInsertQuery = s""" INSERT INTO "public"."Student" (id, name, gpa, gender, \"gradeLevel\", activities) values ('${student.id}','${student.name}','${student.gpa}','${student.gender}','${student.gradeLevel}',$activities)"""
    val result         = Try[Boolean]{
      val connection = new StudentDatabase() getConnection("127.0.0.1", "postgres", "admin")
      val statement  = connection createStatement()
      statement executeUpdate getInsertQuery
      statement close()
      connection close()
      true
    }
    result match {
      case Success(value)     => Future successful value
      case Failure(exception) => Future failed exception
    }
  }

  override def updateData(student: Student): Future[Boolean] = {
    val activities     = student.activities.mkString("Array['", "','", "']")
    val getInsertQuery = s""" UPDATE "public"."Student" SET id = '${student.id}', name = '${student.name}' , gpa = '${student.gpa}' , gender = '${student.gender}', "gradeLevel" = '${student.gradeLevel}', activities = $activities WHERE id = '${student.id}'"""
    val result         = Try[Boolean]{
      val connection = new StudentDatabase() getConnection("127.0.0.1", "postgres", "admin")
      val statement  = connection createStatement()
      statement executeUpdate getInsertQuery
      statement close()
      connection close()
      true
    }

    result match {
      case Success(value)     => Future successful value
      case Failure(exception) => Future failed exception
    }
  }


  override def deleteData(id: AnyVal): Future[Boolean] = {
    val getInsertQuery = s""" DELETE FROM "public"."Student" WHERE id = '$id'"""
    val result         = Try[Boolean]{
      val connection = new StudentDatabase() getConnection("127.0.0.1", "postgres", "admin")
      val statement  = connection createStatement()
      statement executeUpdate getInsertQuery
      statement close()
      connection close()
      true
    }
    result match {
      case Success(value)     => Future successful value
      case Failure(exception) => Future failed exception
    }
  }
}

